### PR TITLE
Contain training run artifact paths

### LINKED ISSUE #1498: M-Courtyard roadmap: P2.1 Training Parameter Safety and Queueing
## Parent
#1497

## What to build
Add explicit trainability checks and a durable admission path for local training so unsupported runs fail early and queued runs survive navigation or process restart.

## Acceptance criteria
- [ ] The plan defines unsupported training configurations before worker launch.
- [ ] The plan defines durable queue state and restore semantics.
- [ ] The unit issues cover both trainability validation and queue admission behavior.
- [ ] The plan records preflight, admission, restore, and cancellation metrics.

## Blocked by
- #1497

## Governing docs
- `docs/reference-scans/m-courtyard-lessons.md`
- `docs/plans/2026-05-24-m-courtyard-improvement-roadmap.md`

## Metrics and probes
trainability preflight latency, queue admission latency, queue restore latency, cancellation latency, unsupported configuration count

## Agent delegation
Pending delegation from the roadmap PR after issue creation.

### PR DESCRIPTION
**Pull request by Zigfreidish**
## Summary
- Adds a pure `defaultTrainingRunName(modelRef:)` admission helper for repository IDs, absolute local model paths, long names, and traversal-shaped input.
- Routes LoRA training queue defaults and worker output directories through managed, sanitized artifact paths under the configured output root.
- Rejects explicit run directories that escape the managed output root before worker launch with `path_escape_detected=true` and the sanitized candidate name.

Closes #1498.

## Plan or Spec
- `docs/plans/2026-05-24-training-parameter-safety-and-queueing.md`
- `docs/plans/2026-07-02-training-run-artifact-path-containment.md`

## Commands Run
```text
swift test --filter 'LoraTrainingJobStoreTests/localTrainingQueueDerivesContainedRunNamesFromModelReferences|LoraTrainingJobStoreTests/localTrainingQueueRejectsEscapingRunDirectoriesWithTypedEvidence|MelixCLIRunnerTests/loraTrainUsesContainedDefaultOutputDirForLocalModelPaths'
PASS: 3 tests.

swift test --filter 'LoraTrainingJobStoreTests/localTrainingQueuePersistsAdmissionRunningCancellationAndMetrics|LoraTrainingJobStoreTests/localTrainingQueueRecordsNonexclusiveDefaultsFailuresAndTerminalCancelErrors|LoraTrainingJobStoreTests/localTrainingQueueKeepsStatusUnchangedWhenCancelRequestCannotBeWritten|MelixCLIRunnerTests/loraTrainPersistsDurableQueueAdmissionBeforeWorkerLaunch|MelixCLIRunnerTests/loraTrainRejectsBusyDurableQueueBeforeWorkerLaunch|MelixCLIRunnerTests/loraRun'
PASS: 9 tests.

swift test --filter LoraTrainingJobStoreTests
PASS: 19 tests.

swift test --filter 'MelixCLIRunnerTests/loraTrain|MelixCLIRunnerTests/loraRun'
PASS: 14 tests.

swift test --filter 'LoraTrainingJobStoreTests/localTrainingQueueDerivesContainedRunNamesFromModelReferences|LoraTrainingJobStoreTests/localTrainingQueueContainmentAcceptsAbsoluteRunPathsUnderFilesystemRoot'
RED: failed before the review fix on whitespace-trimmed traversal and filesystem-root containment.

swift test --filter 'LoraTrainingJobStoreTests/localTrainingQueueDerivesContainedRunNamesFromModelReferences|LoraTrainingJobStoreTests/localTrainingQueueContainmentAcceptsAbsoluteRunPathsUnderFilesystemRoot'
PASS: 2 tests after the review fix.

swift test --filter LoraTrainingJobStoreTests
PASS: 20 tests.

swift test --filter 'MelixCLIRunnerTests/loraTrain|MelixCLIRunnerTests/loraRun'
PASS: 14 tests after the review fix.

git diff --check
PASS.

git diff --check HEAD~1 HEAD
PASS.

python3 scripts/validate_pr_evidence.py --body-file .runtime/pr-bodies/issue-1498.md
PASS: PR evidence looks valid.

git commit -m "fix: contain training run artifact paths"
PASS: pre-commit hook completed and created commit 4bb18bfa. The hook ran make swift-test, make py-test, make integration-test, scoped changed-scope verification, and PR-scoped performance reporting.

git commit -m "fix: handle training path review edge cases"
PASS: pre-commit hook completed and created commit df1796cb. The hook ran make swift-test, make py-test, make integration-test, and PR-scoped performance reporting.
make swift-test: PASS.
make py-test: PASS, 4587 passed, 14 skipped, 2 warnings.
make integration-test: PASS, 123 passed, 1 skipped.
```

## Coverage and Metrics
- Changed-scope coverage: `PASS` via the pre-commit PR-scoped performance row (`coverage_ok=true`).
- Metrics report: `.runtime/pre-commit-performance/20260702-090627-3a07d9aa/report/report.md`
- Report status: `ok`; changed files: `6`; selected probes: `1`; direct/gated probes: `1`; regressions: `0`; context regressions: `0`; verification failures: `0`.
- Probe: `Swift CLI JSON envelope encoding`; targeted tests: `pass`; coverage: `pass`.
- `elapsed_ms_mean`: base `127022.668`, head `13961.906`, delta `-113060.762` (`-89.01%`), status `improvement`.
- Review-fix metrics report: `.runtime/pre-commit-performance/20260702-094656-4bb18bfa/report/report.md`
- Review-fix report status: `ok`; changed files: `2`; selected probes: `0`; regressions: `0`; context regressions: `0`; verification failures: `0`.
- Observability mode: `minimal`; no runtime observability behavior changed.
- Probe overhead: `N/A`; this change uses existing admission tests and existing PR-scoped performance probes.
- Protocol/dependency artifacts: `N/A`; this change does not modify schemas or dependencies.

## Known Gaps
- None.

## Evidence Checklist
- [x] Relevant plan or spec is identified.
- [x] Behavior changes are reflected in the relevant docs.
- [x] Protocol changes include regenerated generated artifacts.
- [x] Dependency changes include updated lockfiles.
- [x] Relevant tests were run.
- [x] A metrics report is included, or `N/A` is stated explicitly with the reason.
- [x] Observability mode, probe overhead, and any deferred debug or evidence work are recorded, or `N/A` is stated explicitly with the reason.
- [x] Deferred work and known gaps are stated explicitly.

### CODE REVIEW

File: Sources/MelixCLICore/LocalTrainingQueueStore.swift [hunk 1]
```diff
@@ -539,6 +591,69 @@ public final class LocalTrainingQueueStore: @unchecked Sendable {
...
+            hash &*= 1_099_511_628_211
+        }
+        return String(format: "%016llx", hash)
+    }
+
+    private static func isPath(_ candidate: URL, containedBy root: URL) -> Bool {
+        let candidatePath = candidate.standardizedFileURL.path
+        let rootPath = root.standardizedFileURL.path
+        return candidatePath == rootPath || candidatePath.hasPrefix(rootPath + "/")
+    }
```
[the remark this answers is not in the document]
Comment (Zigfreidish) [reply Sources/MelixCLICore/LocalTrainingQueueStore.swift:661]: Fixed in df1796cb. `isPath` now uses a suffix-aware root prefix, so a managed output root of `/` accepts absolute child paths instead of checking for `//`. Added `localTrainingQueueContainmentAcceptsAbsoluteRunPathsUnderFilesystemRoot` as the regression test.

File: Sources/MelixCLICore/LocalTrainingQueueStore.swift [hunk 2]
```diff
@@ -539,6 +591,69 @@ public final class LocalTrainingQueueStore: @unchecked Sendable {
...
+        let safeComponents = rawComponents.filter { component in
+            let normalized = component.trimmingCharacters(in: .whitespacesAndNewlines)
+            return !normalized.isEmpty && normalized != "." && normalized != ".."
+        }
+        guard !safeComponents.isEmpty else {
+            return "model"
+        }
+        if rawComponents.contains(where: { $0 == "." || $0 == ".." }) {
+            return safeComponents.last ?? "model"
+        }
```
[the remark this answers is not in the document]
Comment (Zigfreidish) [reply Sources/MelixCLICore/LocalTrainingQueueStore.swift]: Fixed in df1796cb. Traversal detection now trims each raw component before checking `.` and `..`, matching the safe-component filtering behavior. Added the `repo/ .. /Qwen 3.5` regression expectation.
